src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcBase.cs

```diff
@@ -802,6 +829,40 @@ public void Close()
                 _terminateDelegate(_nodeId);
             }
 
+            /// <summary>
+            /// Waits for the child node process to exit.
+            /// </summary>
+            public async Task WaitForExitAsync()
+            {
+                // Wait for the process to exit.
+                if (_isExiting)
+                {
+                    CommunicationsUtilities.Trace("Waiting for node with pid = {0} to exit", _process.Id);
+
+                    // .NET 5 introduces a real WaitForExitAsyc.
+                    // This is a poor man's implementation that uses polling.
+                    int timeout = TimeoutForWaitForExit;
+                    int delay = 5;
+                    while (timeout > 0)
+                    {
+                        bool exited = _process.WaitForExit(milliseconds: 0);
+                        if (exited)
+                        {
+                            return;
+                        }
+                        timeout -= delay;
+                        await Task.Delay(delay).ConfigureAwait(false);
+
+                        // Double delay up to 500ms.
+                        delay = Math.Min(delay * 2, 500);
+                    }
+                }
+
+                // Kill the child and do a blocking wait.
+                CommunicationsUtilities.Trace("Killing node with pid = {0}", _process.Id);
```

Comment:
@cdmihai this message is printed when the node is killed forcefully. I think this addresses your request in https://github.com/dotnet/msbuild/pull/6053#issuecomment-769340348?

Reply:
The node communication tracing logs are pretty obscure and I doubt many users know about them. I was thinking that a LoggingService based warning that actually pops up in the terminal would make the event visible to users so they can investigate or open issues against MSBuild. I find this is the appropriate level of visibility since I think nonresponsive nodes is buggy behaviour on either MSBuild's or the task's part (if there's a long running task that cannot be preempted). I could also be wrong about this requirement (it's not like anyone documented these requirements, who does that? :P). We'll discuss this at our next PR review meeting and let you know.